Class/Practise 2.2/tests.py

```diff
@@ -0,0 +1,135 @@
+import subprocess
+import os
+import sys
+import filecmp
+
+def run_test(executable, input_file, favorite_color, expected_output_file, actual_output_file, expected_console_output, expect_fail=False):
+    print(f"Running test with input file: {input_file}")
+    process = subprocess.Popen([executable], stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
+    
+    # Simulate user input
+    user_input = f"{input_file}\n{favorite_color}\n{actual_output_file}\n"
+    stdout, stderr = process.communicate(user_input)
+
+    if expect_fail:
+        # If we expect the test to fail, check stdout for error messages
+        if stdout.strip() == expected_console_output:
+            print(f"Expected error received: {stdout.strip()}")
+            return True
+        else:
+            print("Expected failure but got none or unexpected error.")
+            print(f"Expected: {expected_console_output}")
+            print(f"Got: {stdout.strip()}")
+            return False
+    else:
+        if stderr:
+            print(f"Unexpected error: {stderr.strip()}")
+            return False
+
+        # Check if the actual output file was created
+        if not os.path.exists(actual_output_file):
+            print(f"Actual output file not found: {actual_output_file}")
+            return False
+
+        # Compare actual output file with expected output file
+        result = filecmp.cmp(expected_output_file, actual_output_file, shallow=False)
+        print(f"Comparison result for {actual_output_file}: {result}")
+        
+        # Compare stdout with expected console output
+        if stdout.strip() != expected_console_output.strip():
+            print(f"Expected console output not received. Got:\n{stdout.strip()}")
+            return False
+        
+        return result
+
+def test_case_1(executable):
```

Comment:
looks good in general
You can add some comments for test cases to describe them briefly like FavoriteColorTopLeftCorner, FavoriteColorInTheMiddle and so on. It may improve readability

Reply:
Added descriptive test case names